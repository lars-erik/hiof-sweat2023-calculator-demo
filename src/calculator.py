import sys

from ExpressionParser import ExpressionParser
from ExpressionReader import ExpressionReader
from Operators import Subtraction, Addition


class CalculatorProgram():

    def __init__(self, operator, parser):
        self.parser = parser
        self.operator = operator

    def read_transform_calculate_format(self):
        operations = self.parser.parse_operations()
        total = self.calculate(operations)
        report = self.format_report(operations, total)
        return report

    def format_report(self, operations, total):
        report = ''
        for op in operations:
            report += f"{op.left} {str(op.operator)} {op.right} = {op.calculate()}\n"
        report += 'Total: ' + str(total)
        return report

    def calculate(self, operations):
        total = 0
        for op in operations:
            subtotal = op.calculate()
            total = self.operator.calculate(total, subtotal)
        return total



if __name__ == "__main__":
    operator = (
        Subtraction() if len(sys.argv) > 1 and sys.argv[1] == 'sub'
        else
        Addition()
    )
    parser = ExpressionParser(ExpressionReader(), operator)
    program = CalculatorProgram(operator, parser)

    print(program.read_transform_calculate_format())
