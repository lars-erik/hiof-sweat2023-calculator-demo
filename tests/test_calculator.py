from approvaltests import approvals

from calculator import read_transform_calculate_format


def test_calculations():
   report = read_transform_calculate_format()
   approvals.verify(report)

