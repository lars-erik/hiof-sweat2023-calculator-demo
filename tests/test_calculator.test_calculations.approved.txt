1 + 1 = 2
1 + 2 = 3
5 + 10 = 15
1205 + 795 = 2000
Total: 2020
